umbral/keys.py

```diff
@@ -239,3 +261,61 @@ def __eq__(self, other):
         else:
             is_eq = False
         return is_eq
+
+
+class UmbralKeyingMaterial(AbstractUmbralKeyingMaterial):
+    """
+    This class handles keying material for Umbral, by allowing deterministic
+    derivation of UmbralPrivateKeys based on labels. 
+    Don't use this key material directly as a key.
+    
+    """
+
+    def __init__(self, keying_material: bytes=None):
+        """
+        Initializes an UmbralKeyingMaterial.
+        """
+        if keying_material:
+            if len(keying_material) < 32:
+                raise ValueError("UmbralKeyingMaterial must have size at least 32 bytes.")
+            self.keying_material = keying_material
+        else:
+            self.keying_material = os.urandom(32)
+
+    def _get_keying_material(self):
+        return self.keying_material
+
+    def derive_private_key_from_label(self, label: bytes, 
+                                    salt: bytes=None, params: UmbralParameters=None):
+        """
+        Derives an UmbralPrivateKey using a KDF from this instance of 
+        UmbralKeyingMaterial, a label, and an optional salt.
+        """
+        if params is None:
+            params = default_params()
+
+        hkdf = HKDF(algorithm=hashes.BLAKE2b(64),
+                    length=64,
+                    salt=salt,
+                    info=b"NuCypherKMS/KeyDerivation/"+label,
+                    backend=default_backend()
+                    )
+
+        bn_key = hash_to_bn(hkdf.derive(self.keying_material), params)
+        return UmbralPrivateKey(bn_key, params)
+
+    @classmethod
+    def from_bytes(cls, key_data: bytes, password: bytes=None, _scrypt_cost: int=20):
```

Comment:
Let's keep `to_bytes` and `from_bytes` in `UmbralPrivateKey` so that it matches with `UmbralPublicKey`.

If we want to make a move to put these in a base class, we need to think more about structuring the rest of the key classes like that. Either way, I think something like this should be in its own independent PR.